feat(next): cleanup tasks

StaberindeZA opened the pull request:
## Because

* During development some console errors and warnings were introduced
* Improve developer experience and code quality of payments next

## This pull request

* Update imports in Payments Next from root project directory
* Resolve browser console errors
  * Added missing CSP content items
  * Removed Layout image width to remove next/image warning
* Resolved server console error related to nest-typed-config
* Extract CheckoutForm button into PrimaryButton component
* Added Stripe Public API key config variable
* Updated checkout page headings
* Updated Stripe Payment Element to use accordion

## Issue that this pull request solves

Closes: #FXA-9625

## Checklist

_Put an `x` in the boxes that apply_

- [x] My commit is GPG signed.
- [x] If applicable, I have modified or added tests which pass locally.
- [x] I have added necessary documentation (if appropriate).
- [x] I have verified that my changes render correctly in RTL (if appropriate).

sardesam commented on apps/payments/next/.env.production:
```diff
@@ -24,9 +24,13 @@ MYSQL_CONFIG__ACQUIRE_TIMEOUT_MILLIS=
 
+# Stripe
+STRIPE_PUBLIC_API_KEY=
+
```
Nit: similar to how you added `# StripeConfig` on L27, could you add `# PayPal Config`?

sardesam replied:
Nit: I think this was missed.

sardesam replied:
I think this was missed.

sardesam commented on apps/payments/next/app/[locale]/[offeringId]/checkout/[interval]/[cartId]/layout.tsx:
```diff
@@ -13,6 +13,7 @@ import {
 } from '@fxa/payments/ui/server';
 import { getFakeCartData } from '../../../../../_lib/apiClient';
```
Suggestion: rename to `apps/payments/next/app/_lib/apiClient` similar to how it's written for `apps/payments/next/app/[locale]/[offeringId]/checkout/[interval]/[cartId]/success/page.tsx` L11.

pdehaan replied:
Good catch!
Looks like this _might_ be the only one:

```sh
git grep -Pin "\/\.\.\/_lib\/" apps libs | cat

apps/payments/next/app/[locale]/[offeringId]/checkout/[interval]/[cartId]/layout.tsx:14:import { getFakeCartData } from '../../../../../_lib/apiClient';
```

versus

```sh
git grep -Pin "\/\w+\/_lib\/" apps libs | cat

apps/payments/next/app/[locale]/[offeringId]/checkout/[interval]/[cartId]/start/page.tsx:16:} from 'apps/payments/next/app/_lib/apiClient';
apps/payments/next/app/[locale]/[offeringId]/checkout/[interval]/[cartId]/success/page.tsx:11:import { getFakeCartData } from 'apps/payments/next/app/_lib/apiClient';
```

sardesam commented on apps/payments/next/middleware.ts:
```diff
@@ -4,6 +4,17 @@
 import { NextRequest, NextResponse } from 'next/server';
 
 export function middleware(request: NextRequest) {
+  // Read env vars directly from process.env
+  // Currently not possible to use app/config in middleware
+  const accountsStaticCdn = process.env.CSP__ACCOUNTS_STATIC_CDN;
```
Nit: Could you add a link or a 'As of [date]' so when we visit in future we'll have a better idea of the status and whether it will become possible.

sardesam commented on apps/payments/next/middleware.ts:
```diff
@@ -4,6 +4,17 @@
+   *  - Next.js next/image currently causes an inline style CSP error.
+   *    There is a work around available, however at this time, we've opted
+   *    to use 'unsafe-inline' to match what's in fxa-payments-server
+   *    https://github.com/vercel/next.js/issues/45184
```
Nice. Thank you for the reference! 👍

sardesam commented on libs/payments/ui/src/lib/client/components/PaymentSection.tsx:
```diff
@@ -10,6 +10,9 @@ import { StripeWrapper } from './StripeWrapper';
 import { Localized } from '@fluent/react';
 
 interface PaymentFormProps {
+  config: {
+    stripePublicApiKey: string;
+  };
```
Q: Out of curiosity, how often do we have to pass config information in the client components? Wondering if we could use a context provider so all child components will already have access to it?

StaberindeZA replied:
Oh yeah, that's a good idea! Let me put that in.

sardesam commented on libs/payments/ui/src/lib/server/purchase-details.tsx:
```diff
@@ -6,7 +6,6 @@ import { Invoice } from '@fxa/payments/cart';
 import Image from 'next/image';
 import { formatPlanPricing } from '../utils/helpers';
```
Nit: change to `@fxa/payments/ui`

StaberindeZA replied:
Since `formatPlanPricing` is in the same library, relative path here makes sense.

sardesam reviewed (commented):
Looking good! Left a few comments and had a question (see inline comment below).

pdehaan commented on apps/payments/next/app/[locale]/[offeringId]/checkout/[interval]/[cartId]/error/page.tsx:
```diff
@@ -8,7 +8,7 @@ import Link from 'next/link';
 import { DEFAULT_LOCALE } from '@fxa/shared/l10n';
 
-import errorIcon from '../../../../../../../images/error.svg';
+import errorIcon from '@fxa/shared/assets/images/error.svg';
```
😍

pdehaan commented on apps/payments/next/app/[locale]/[offeringId]/checkout/[interval]/[cartId]/start/page.tsx:
```diff
@@ -47,12 +48,12 @@ export default async function Checkout({ params }: { params: CheckoutParams }) {
       {!session && (
         <>
-          <h4 className="font-semibold text-grey-600 text-lg mt-10">
+          <h2 className="font-semibold text-grey-600 text-lg mt-10">
```
Not sure it matters, but the styling on this `<h2>` is _slightly_ different than the 2 below:

```sh
git grep -n -A1 "<h2" -- apps | grep "className=[\"'].*[\"']"

apps/payments/next/app/[locale]/[offeringId]/checkout/[interval]/[cartId]/start/page.tsx:51:          <h2 className="font-semibold text-grey-600 text-lg mt-10">
apps/payments/next/app/[locale]/[offeringId]/checkout/[interval]/[cartId]/start/page.tsx-98-              className="font-semibold text-grey-600 text-lg mt-14 mb-5"
apps/payments/next/app/[locale]/[offeringId]/checkout/[interval]/[cartId]/start/page.tsx-108-             className="font-semibold text-grey-600 text-lg mt-14 mb-5"
```

```diff
+ mt-10
- mt-14 mb-5
```

StaberindeZA replied:
This was copied from existing payments-server. I'd rather keep it as is for now.

pdehaan commented on libs/payments/ui/src/lib/client/components/CheckoutForm.tsx:
```diff
@@ -162,25 +166,31 @@ export function CheckoutForm({ readOnly, cart }: CheckoutFormProps) {
+              className="h-4 w-4 my-0 mx-3 relative top-0.5"
+              alt=""
+            />
+            Subscribe Now
```
Does this need to be l10n'ed? (I see we're translating the `placeholder=` above on L147).

StaberindeZA replied:
Ah good catch thank you. Deleted too much when I added PrimaryButton.

pdehaan commented on libs/payments/ui/src/lib/client/components/PrimaryButton.tsx:
```diff
@@ -0,0 +1,15 @@
+interface PrimaryButtonProps
```
```suggestion
/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

interface PrimaryButtonProps
```

pdehaan commented on libs/payments/ui/src/lib/client/components/StripeWrapper.tsx:
```diff
@@ -19,21 +22,21 @@ interface StripeWrapperProps {
-  const stripePromise = loadStripe(
-    'pk_test_VNpCidC0a2TJJB3wqXq7drhN00sF8r9mhs'
-  );
+  const stripePromise = loadStripe(config.stripePublicApiKey);
```
nit: Do we still need the `TODO` above on L31?

pdehaan commented on package.json:
```diff
@@ -258,6 +258,7 @@
     "minimist": "^1.2.6",
     "moment:>2.0.0 <3": ">=2.29.4",
     "node-forge": ">=1.3.0",
+    "nest-typed-config/class-validator": "0.14.1",
```
<s>Is this used somewhere? I don't see any file in this diff using it, or in this `git grep` (unless we oddly need it as a peer dep):

```sh
git grep "nest-typed-config/class-validator" | cat

package.json:    "nest-typed-config/class-validator": "0.14.1",
```
</s>

Nevermind, I see it's in `resolutions` and not as a `dependency` or `devDependency`.

StaberindeZA commented on apps/payments/next/.env.development:
```diff
@@ -32,6 +32,9 @@ GEODB_MANAGER_CONFIG__LOCATION_OVERRIDE__POSTAL_CODE=11211
 STRIPE_CONFIG__TAX_IDS={}
 
+# Stripe
+STRIPE_PUBLIC_API_KEY=pk_test_VNpCidC0a2TJJB3wqXq7drhN00sF8r9mhs
```
This is the Stripe public API key for the Stripe SUBPLAT_DEV environment. It's fine to share publicly, and we currently have it as default in `fxa-payments-server` as well.

StaberindeZA commented on apps/payments/next/.env.development:
```diff
@@ -63,3 +66,9 @@ STATS_D_CONFIG__MAX_BUFFER_SIZE=
+CSP__ACCOUNTS_STATIC_CDN=https://accounts-static.cdn.mozilla.net
+
+# Other
+CONTENT_SERVER_URL=https://accounts.stage.mozaws.net
```
Oh yeah localhost probably makes more sense.

pdehaan commented on apps/payments/next/config/index.ts:
```diff
@@ -1,11 +1,17 @@
+}
+
 class AuthJSConfig {
   @IsString()
```
Unrelated, but should this and L19-20 use this new `@IsUrl()` decorator too?

```diff
-  @IsString()
+  @IsUrl()
  issuerUrl!: string;

-  @IsString()
+  @IsUrl()
  wellKnownUrl!: string;
```

sardesam replied:
^ Good question. Curious about this too.

StaberindeZA replied:
Yup makes sense.

pdehaan replied:
This is very hard to grep, apparently, but this might be an assortment of unreliable results where we're using `@IsString()` that could be converted to `@IsUrl()`:
```sh
git grep -A3 "@IsString()" | grep -E "( string)" | grep -Ei "(Url|Uri|redirectTo)"

apps/payments/next/config/index.ts-  issuerUrl!: string;
apps/payments/next/config/index.ts-  wellKnownUrl!: string;
packages/fxa-settings/src/models/integrations/client-info.ts-  imageUri: string | undefined;
packages/fxa-settings/src/models/integrations/client-info.ts-  redirectUri: string | undefined;
packages/fxa-settings/src/models/integrations/oauth-integration.ts-  imageUri: string | undefined;
packages/fxa-settings/src/models/integrations/oauth-integration.ts-  declare redirectTo: string | undefined;
packages/fxa-settings/src/models/integrations/oauth-integration.ts-  redirectUrl: string | undefined;
packages/fxa-settings/src/models/integrations/oauth-integration.ts-  redirectUri: string | undefined;
packages/fxa-settings/src/models/integrations/signin-signup-info.ts-  redirectUri: string | undefined;
packages/fxa-settings/src/models/integrations/signin-signup-info.ts-  redirectTo: string | undefined;
packages/fxa-settings/src/models/integrations/web-integration.ts-  redirectTo: string | undefined;
packages/fxa-settings/src/models/pages/signin/query-params.ts-  redirectTo: string | undefined = undefined;
```

Although guessing everything in fxa-settings is out of scope here.

sardesam reviewed (commented):
Re-reviewed and updates look good. Thank you. Few inline questions that are non-blockers.

I had one question about config - do we still need to explicitly pass it in `ToS` and `PaymentForm` now that `ConfigProvider` was added?

StaberindeZA commented:
> Thoughts on adding the italicized lines below as headings?
> 
>     * h1 - Subscription Title
>       
>       * h2 - Product Name
>         
>         * h3 - Product Details
>         * _h3 - Total_
>       * h2 - 1. Create Mozilla account
>       * h2 - 2. Choose your payment method
>         
>         * _h3 - First you'll need to approve your subscription_
>         * _h3 - Enter your card information_
>         * h3 - Firefox Accounts TOS [1][2]
>         * h3 - Product TOS [1]
> 
> 
> [1] Open to a later discussion on whether these should be included as headings [2] Will this eventually be updated to Mozilla accounts?

Made the suggested heading updates. 

For [1] I'd like to keep as is for now, since we are potentially looking at adopting an actual Ts&Cs footer, similar to settings.
For [2], I've updated to Mozilla Accounts

StaberindeZA commented:
> Re-reviewed and updates look good. Thank you. Few inline questions that are non-blockers.
> 
> I had one question about config - do we still need to explicitly pass it in `ToS` and `PaymentForm` now that `ConfigProvider` was added?

ToS is a server component so it won't have access to the App Context. I've updated ToS though to remove the "config" verbiage, and just use plane `contentServerUrl` so that it's more clear that it's just a regular prop required by the Component.

Payment Section, ah yes good point. I moved Providers outside of PaymentSection, so now it's no longer necessary in PaymentSection. Note that Providers will most likely move again once  https://github.com/mozilla/fxa/pull/16960 lands

sardesam reviewed (approved):
Thank you for making those updates and addressing the questions!

pdehaan commented on apps/payments/next/.env.production:
```diff
@@ -1,11 +1,11 @@
 # Auth
-AUTH__ISSUER_URL=
-AUTH__WELL_KNOWN_URL=
+AUTH__ISSUER_URL=http://localhost:3030
+AUTH__WELL_KNOWN_URL=http://localhost:3030/.well-known/openid-configuration
```
Q: Should the .env.production env file have localhost domains? Or are those _for_ production since they're local to the k8s pods?

sardesam commented:
> > Re-reviewed and updates look good. Thank you. Few inline questions that are non-blockers.
> > I had one question about config - do we still need to explicitly pass it in `ToS` and `PaymentForm` now that `ConfigProvider` was added?
> 
> ToS is a server component so it won't have access to the App Context. I've updated ToS though to remove the "config" verbiage, and just use plane `contentServerUrl` so that it's more clear that it's just a regular prop required by the Component.
> 
> Payment Section, ah yes good point. I moved Providers outside of PaymentSection, so now it's no longer necessary in PaymentSection. Note that Providers will most likely move again once #16960 lands

Ah, I forgot ToS is a server component. Thank you for making the change to PaymentSection.

Noted re: Providers move 👍
